kv_storage.h

```diff
@@ -0,0 +1,383 @@
+struct SpecificEntryTypeKeyNotFoundException : KeyNotFoundException {
+  typedef typename std::remove_reference<decltype(std::declval<T_ENTRY>().key())>::type T_KEY;
+  T_KEY key;
+  SpecificEntryTypeKeyNotFoundException(const T_KEY& key) : key(key) {}
+};
+
+struct KeyAlreadyExistsException : bricks::Exception {};
+
+template <typename T_ENTRY>
+struct SpecificEntryKeyAlreadyExistsException : KeyAlreadyExistsException {
+  T_ENTRY entry;
+  SpecificEntryKeyAlreadyExistsException(const T_ENTRY& entry): entry(entry) {}
+};
+
+struct EntryShouldExistException : bricks::Exception {};
+
+template <typename T_ENTRY>
+struct SpecificEntryShouldExistException : EntryShouldExistException {
+  T_ENTRY entry;
+  SpecificEntryShouldExistException(const T_ENTRY& entry): entry(entry) {}
+};
+
+// API calls, `[Async]{Get/Add/Update/Delete}(...)`.
+//
+// C++ API supports three syntaxes:
+//
+//   1) Asynchronous, using std::{promise/future}.
+//      To use the async/await paradigm.
+//      `AsyncAdd(entry)` and `AsyncGet(key)`, both return a future. The `AsyncGet()` one may throw.
+//
+//   2) Asynchronous, using std::function as a callback.
+//      Best for HTTP endpoints, with the Request object passed into this callback.
+//      `AsyncAdd(entry, on_done, on_exists)` and `AsyncGet(key, on_found, on_not_found)`, both return void.
+//
+//   3) Synchronous. `Get(key)` and `Add(entry)`. The `Get()` one may throw.
+
+// TODO(dkorolev): Polymorphic types. Pass them in as std::tuple<...>, or directly as a variadic template param.
+
+// TODO(dkorolev): How about HTTP endpoints? They should be added somewhat automatically, right?
+//                 And it would require those `Scoped*` endpoints in Bricks. On me.
+
+// TODO(dkorolev): So, are we CRUD+REST or just a KeyValueStorage? Now we kinda merged POST+PUT. Not cool.
+
+// Policies.
+template <typename ENTRY>
+constexpr auto has_nonthrowing_get(int) -> decltype(std::declval<ENTRY>().allow_nonthrowing_get, bool()) {
+  return true;
+}
+
+template <typename ENTRY>
+constexpr bool has_nonthrowing_get(...) {
+  return false;
+}
+
+template <typename ENTRY, bool>
+struct ExtractNonThrowingGet {
+ constexpr static bool value = ENTRY::allow_nonthrowing_get;
+ static_assert(!value || std::is_base_of<Nullable, ENTRY>::value, "Entry type must be derived from Nullable.");
+};
+
+template <typename ENTRY>
+struct ExtractNonThrowingGet<ENTRY, false> {
+  constexpr static bool value = false;  // Default: throw if key not found.
+};
+
+template <typename ENTRY>
+constexpr auto has_overwrite_on_add(int) -> decltype(std::declval<ENTRY>().allow_overwrite_on_add, bool()) {
+  return true;
+}
+
+template <typename ENTRY>
+constexpr bool has_overwrite_on_add(...) {
+  return false;
+}
+
+template <typename ENTRY, bool>
+struct ExtractOverwriteOnAdd {
+ constexpr static bool value = ENTRY::allow_overwrite_on_add;
+ static_assert(!value || std::is_base_of<Nullable, ENTRY>::value, "Entry type must be derived from Nullable.");
```

Comment:
Same comment as above regarding the `static_assert` message. Overall :+1: the idea.